qemu/tests/qemu_guest_agent.py

```diff
@@ -3744,6 +3745,82 @@ def __init__(self, test, params, env):
         self.qemu_ga_pkg = params.get("qemu_ga_pkg", "")
         self.gagent_src_type = params.get("gagent_src_type", "url")
 
+    @error_context.context_aware
+    def _check_serial_driver(self, test, params, env):
+        '''
+        Check whether serial device is installed or not, if not install it.
+        Test steps:
+        1) Check whether serial driver is running.
+        2) Install serial driver.
+        3) Reboot vm
+        4) Verify serial has been installed.
+
+        :param test: QEMU test object
+        :param params: Dictionary with the test parameters
+        :param env: Dictionary with test environment
+        '''
+
+        session = self._get_session(params, self.vm)
+        self._open_session_list.append(session)
+
+        OPERATION_TIMEOUT = 120
+        INSTALL_TIMEOUT = 360
+        driver_name = params["driver_name"]
+        device_name = params["device_name"]
+        device_hwid = params["device_hwid"]
+        media_type = params["virtio_win_media_type"]
+
+        try:
+            get_drive_letter = getattr(virtio_win,
+                                       "drive_letter_%s" % media_type)
+            get_product_dirname = getattr(virtio_win,
+                                          "product_dirname_%s" % media_type)
+            get_arch_dirname = getattr(virtio_win,
+                                       "arch_dirname_%s" % media_type)
+        except AttributeError:
+            test.error("Not supported virtio win media type '%s'",
+                       media_type)
+        viowin_ltr = get_drive_letter(session)
+        if not viowin_ltr:
+            test.error("Could not find virtio-win drive in guest")
+        guest_name = get_product_dirname(session)
+        if not guest_name:
+            test.error("Could not get product dirname of the vm")
+        guest_arch = get_arch_dirname(session)
+        if not guest_arch:
+            test.error("Could not get architecture dirname of the vm")
+        inf_middle_path = ("{name}\\{arch}" if media_type == "iso"
+                           else "{arch}\\{name}").format(name=guest_name,
+                                                         arch=guest_arch)
+        inf_find_cmd = 'dir /b /s %s\\%s.inf | findstr "\\%s\\\\"'
+        inf_find_cmd %= (viowin_ltr, driver_name, inf_middle_path)
+        inf_path = session.cmd(inf_find_cmd,
+                               timeout=OPERATION_TIMEOUT).strip()
+        logging.info("Found inf file '%s'", inf_path)
+
+        error_context.context("Check whether serial drive is running.",
+                              logging.info)
+        if params["gagent_serial_type"] == "virtio":
```

Comment:
If this is for the purpose: `only for vioser method`, I think you might control it in cfg file, not necessary to check it again, and also all the above code seems meaningless for isa-serial, please correct me if I was wrong.
And also in the bz I found `only for windows guest`, so from my understanding it might like:
```
diff --git a/qemu/tests/cfg/qemu_guest_agent.cfg b/qemu/tests/cfg/qemu_guest_agent.cfg
index 14033cb5..11a5b11f 100644
--- a/qemu/tests/cfg/qemu_guest_agent.cfg
+++ b/qemu/tests/cfg/qemu_guest_agent.cfg
@@ -159,9 +159,11 @@
             format_disk_cmd = "yes|mkfs.ext4 DISK"
             write_disk_cmd = "dd if=/dev/zero of=${mount_point}/file bs=1M count=800 oflag=direct"
             delete_file_cmd = "rm ${mount_point}/file -f; sync"
+            check_vioser = no
             Windows:
                 # for windows guest, only support win8+
                 no Win7,Win2008
+                check_vioser = yes
                 start_vm = yes
                 driver_name = vioscsi
                 images += " stg"
@@ -447,12 +449,14 @@
             gagent_check_type = qgastatus_after_remove_qga
     variants:
         - virtio_serial:
+            check_vioser = yes
```